Elasticsearch error in top-vulnerable-nodes api

```
[2023-02-02 22:03:13 +0000] [240] [ERROR] Exception on /deepfence/v1.5/vulnerability/top_vulnerable_nodes [GET]
Traceback (most recent call last):

File "/app/code/api/vulnerability_api.py", line 1416, in get_top_vulnerable_nodes
  aggs_responses = ESConn.msearch(search_queries).get("responses", [])

elasticsearch.exceptions.TransportError: TransportError(429, 'circuit_breaking_exception', '[parent] Data too large, data for [<http_request>] would be [1527533752/1.4gb], which is larger than the limit of [1513521152/1.4gb], real usage: [1527416216/1.4gb], new bytes reserved: [117536/114.7kb], usages [request=0/0b, fielddata=172677/168.6kb, in_flight_requests=117536/114.7kb, accounting=1144270/1mb]')
```

--- a/deepfence_backend/api/vulnerability_api.py
+++ b/deepfence_backend/api/vulnerability_api.py
@@ -1363,10 +1363,12 @@ def get_top_vulnerable_nodes():
     recent_image_scan_ids = get_recent_scan_ids(CVE_SCAN_LOGS_INDEX, number, time_unit, None,
                                                {"action": "COMPLETED", "node_id": active_images})[:ES_MAX_CLAUSE]
     response = {NODE_TYPE_HOST: [], NODE_TYPE_CONTAINER_IMAGE: []}
-    if not recent_host_scan_ids or not recent_image_scan_ids:
+    if not recent_host_scan_ids and not recent_image_scan_ids:
         return set_response(data=response)
 
     def get_top_vulnerability_counts(scan_ids):
+        if not scan_ids:
+            return []
         aggs = {
             "cve_container_image": {
                 "terms": {

--- a/deepfence_backend/api/vulnerability_api.py
+++ b/deepfence_backend/api/vulnerability_api.py
@@ -9,7 +9,7 @@ from datetime import datetime
 import urllib.parse
 import magic
 from utils.helper import websocketio_channel_name_format, is_network_attack_vector, get_image_cve_status, \
-    set_vulnerability_status_for_packages, get_top_exploitable_vulnerabilities
+    set_vulnerability_status_for_packages, get_top_exploitable_vulnerabilities, get_recent_scan_ids
 from utils.response import set_response, format_response
 from utils.common import calculate_interval, format_es_resp_for_api
 from utils.custom_exception import InvalidUsage, InternalError, DFError
@@ -1358,72 +1358,47 @@ def get_top_vulnerable_nodes():
                     not node_details.get("pseudo", False) and node_details.get("docker_container_state") == "running":
                 if node_details["image_name_with_tag"] not in active_images:
                     active_images.append(node_details["image_name_with_tag"])
-    # TODO: es max clause
-    active_hosts = active_hosts[:ES_MAX_CLAUSE]
-    active_images = active_images[:ES_MAX_CLAUSE]
-    aggs = {
-        "cve_container_image": {
-            "terms": {
-                "field": "cve_container_image.keyword",
-                "size": 5
-            },
-            "aggs": {
-                "scan_id": {
-                    "terms": {
-                        "field": "scan_id.keyword",
-                        "size": 1,
-                        "order": {"recent_scan_aggr": "desc"}
-                    },
-                    "aggs": {
-                        "recent_scan_aggr": {
-                            "max": {
-                                "field": "@timestamp"
-                            }
-                        },
-                        "cve_severity": {
-                            "terms": {
-                                "field": "cve_severity.keyword",
-                                "size": ES_TERMS_AGGR_SIZE
-                            }
+    recent_host_scan_ids = get_recent_scan_ids(CVE_SCAN_LOGS_INDEX, number, time_unit, None,
+                                          {"action": "COMPLETED", "node_id": active_hosts})[:ES_MAX_CLAUSE]
+    recent_image_scan_ids = get_recent_scan_ids(CVE_SCAN_LOGS_INDEX, number, time_unit, None,
+                                               {"action": "COMPLETED", "node_id": active_images})[:ES_MAX_CLAUSE]
+    response = {NODE_TYPE_HOST: [], NODE_TYPE_CONTAINER_IMAGE: []}
+    if not recent_host_scan_ids or not recent_image_scan_ids:
+        return set_response(data=response)
+
+    def get_top_vulnerability_counts(scan_ids):
+        aggs = {
+            "cve_container_image": {
+                "terms": {
+                    "field": "cve_container_image.keyword",
+                    "size": 5
+                },
+                "aggs": {
+                    "cve_severity": {
+                        "terms": {
+                            "field": "cve_severity.keyword",
+                            "size": ES_TERMS_AGGR_SIZE
                         }
                     }
                 }
             }
         }
-    }
-    host_filters = {"node_type": NODE_TYPE_HOST, "cve_container_image": active_hosts}
-    host_query = ESConn.aggregation_helper(CVE_INDEX, host_filters, aggs, number,
-                                           TIME_UNIT_MAPPING.get(time_unit), lucene_query_string, get_only_query=True)
-    container_filters = {"node_type": NODE_TYPE_CONTAINER_IMAGE, "cve_container_image": active_images}
-    container_query = ESConn.aggregation_helper(CVE_INDEX, container_filters, aggs, number,
-                                                TIME_UNIT_MAPPING.get(time_unit), lucene_query_string,
-                                                get_only_query=True)
-    search_queries = [
-        {"index": CVE_INDEX}, host_query,
-        {"index": CVE_INDEX}, container_query
-    ]
-    aggs_responses = ESConn.msearch(search_queries).get("responses", [])
-    response = {NODE_TYPE_HOST: [], NODE_TYPE_CONTAINER_IMAGE: []}
+        filters = {"masked": False, "scan_id": scan_ids}
+        aggs_responses = ESConn.aggregation_helper(CVE_INDEX, filters, aggs, number,
+                                          TIME_UNIT_MAPPING.get(time_unit), lucene_query_string)
+        resp = []
+        for node_aggs in aggs_responses.get("aggregations", {}).get("cve_container_image", {}).get("buckets", []):
+            for severity_bkt in node_aggs["cve_severity"]["buckets"]:
+                node_data = {}
+                node_data["node"] = node_aggs["key"]
+                node_data["type"] = severity_bkt["key"]
+                node_data["value"] = severity_bkt["doc_count"]
+                resp.append(node_data)
+        return resp
 
-    if "aggregations" in aggs_responses[0] and active_hosts:
-        for host_aggs in aggs_responses[0]["aggregations"]["cve_container_image"]["buckets"]:
-            for scan_id_bkt in host_aggs["scan_id"]["buckets"]:
-                for severity_bkt in scan_id_bkt["cve_severity"]["buckets"]:
-                    host_data = {}
-                    host_data["node"] = host_aggs["key"]
-                    host_data["type"] = severity_bkt["key"]
-                    host_data["value"] = severity_bkt["doc_count"]
-                    response[NODE_TYPE_HOST].append(host_data)
+    response[NODE_TYPE_HOST].extend(get_top_vulnerability_counts(recent_host_scan_ids))
+    response[NODE_TYPE_CONTAINER_IMAGE].extend(get_top_vulnerability_counts(recent_image_scan_ids))
 
-    if "aggregations" in aggs_responses[1] and active_images:
-        for image_aggs in aggs_responses[1]["aggregations"]["cve_container_image"]["buckets"]:
-            for scan_id_bkt in image_aggs["scan_id"]["buckets"]:
-                for severity_bkt in scan_id_bkt["cve_severity"]["buckets"]:
-                    image_data = {}
-                    image_data["node"] = image_aggs["key"]
-                    image_data["type"] = severity_bkt["key"]
-                    image_data["value"] = severity_bkt["doc_count"]
-                    response[NODE_TYPE_CONTAINER_IMAGE].append(image_data)
     return set_response(data=response)
 
 
@@ -1542,9 +1517,13 @@ def list_process_open_files():
 
     pattern = '^\((.*?):(.*?)\)$'
 
-    omitted_cve_scan_log_list = ['cve_container_image', 'cve_container_image_id', 'cve_caused_by_package', 'cve_cvss_score', 'cve_type', 'doc_id', 'cve_id','cve_caused_by_package_path', 'cve_id_cve_severity_cve_container_image', 'cve_description', 'cve_container_layer', 'cve_container_name','cve_attack_vector', 'cve_severity', 'cve_overall_score', 'cve_link', 'count', 'cve_fixed_in']
+    omitted_cve_scan_log_list = ['cve_container_image', 'cve_container_image_id', 'cve_caused_by_package',
+                                 'cve_cvss_score', 'cve_type', 'doc_id', 'cve_id', 'cve_caused_by_package_path',
+                                 'cve_id_cve_severity_cve_container_image', 'cve_description', 'cve_container_layer',
+                                 'cve_container_name', 'cve_attack_vector', 'cve_severity', 'cve_overall_score',
+                                 'cve_link', 'count', 'cve_fixed_in']
 
-    omitted_list = ['host_name', 'kubernetes_cluster_name','type', 'host']
+    omitted_list = ['host_name', 'kubernetes_cluster_name', 'type', 'host']
 
     lucene_query_string_cve_scan = deepcopy(lucene_query_string)
     if lucene_query_string_cve_scan:
@@ -1560,12 +1539,10 @@ def list_process_open_files():
             if result.group(1) in omitted_list:
                 lucene_query_string_sbom = ""
 
-
     if lucene_query_string:
         lucene_query_string = urllib.parse.unquote(lucene_query_string)
     else:
         lucene_query_string = ""
-
 
     start_index = request.args.get("start_index", 0)
     size = request.args.get("size", 10)
@@ -1610,14 +1587,12 @@ def list_process_open_files():
         lucene_query_string_cve_scan, add_masked_filter=False, get_only_query=True)
     cve_scan_resp = ESConn.search(CVE_SCAN_LOGS_INDEX, aggs_query_cve_scan, 0, 0)
 
-
     filters = request.args.get("filters", {})
     filters["scan_id"] = []
     if "aggregations" in cve_scan_resp and node_id:
         for node_id_aagr in cve_scan_resp["aggregations"]["node_id"]["buckets"]:
             for scan_id_doc in node_id_aagr["completed_scans"].get("hits", {}).get("hits", []):
                 filters["scan_id"].append(scan_id_doc["_source"]["scan_id"])
-
 
     total_scans = len(filters["scan_id"])
     filters["scan_id"] = filters["scan_id"][slice(int(start_index), int(start_index) + int(size))]
@@ -1647,7 +1622,9 @@ def list_process_open_files():
         }
     }
 
-    sbom_aggs_response = ESConn.aggregation_helper(SBOM_ARTIFACT_INDEX, filters, sbom_aggs, number, TIME_UNIT_MAPPING.get(time_unit), lucene_query_string_sbom, add_masked_filter=False)
+    sbom_aggs_response = ESConn.aggregation_helper(SBOM_ARTIFACT_INDEX, filters, sbom_aggs, number,
+                                                   TIME_UNIT_MAPPING.get(time_unit), lucene_query_string_sbom,
+                                                   add_masked_filter=False)
 
     open_files_map = {}
     if "aggregations" in sbom_aggs_response:
@@ -1692,7 +1669,8 @@ def list_process_open_files():
         open_files_list.append(node_info)
 
     open_files_list_final = []
-    final_res = set_vulnerability_status_for_packages(open_files_list, number, time_unit, lucene_query_string, open_files_map, open_files_list_final)
+    final_res = set_vulnerability_status_for_packages(open_files_list, number, time_unit, lucene_query_string,
+                                                      open_files_map, open_files_list_final)
     return set_response(data={"data": final_res, "total": total_scans})
 
 

